applovin_max/android/src/main/java/com/applovin/applovin_max/AppLovinMAXAdViewPlatformWidget.java

```diff
@@ -0,0 +1,205 @@
+
+    @Nullable
+    private AppLovinMAXAdView containerView;
+
+    public AppLovinMAXAdViewPlatformWidget(final String adUnitId, final MaxAdFormat adFormat, final AppLovinSdk sdk, final Context context)
+    {
+        this( adUnitId, adFormat, false, sdk, context );
+    }
+
+    public AppLovinMAXAdViewPlatformWidget(final String adUnitId, final MaxAdFormat adFormat, final boolean preload, final AppLovinSdk sdk, final Context context)
+    {
+        adView = new MaxAdView( adUnitId, adFormat, sdk, context );
+        adView.setListener( this );
+        adView.setRevenueListener( this );
+
+        adView.setExtraParameter( "adaptive_banner", "true" );
+
+        // Set this extra parameter to work around a SDK bug that ignores calls to stopAutoRefresh()
+        adView.setExtraParameter( "allow_pause_auto_refresh_immediately", "true" );
+
+        isPreloadPlatformWidget = preload;
+    }
+
+    public MaxAdView getAdView()
+    {
+        return adView;
+    }
+
+    public void setPlacement(@Nullable final String value)
+    {
+        adView.setPlacement( value );
+    }
+
+    public void setCustomData(@Nullable final String value)
+    {
+        adView.setCustomData( value );
+    }
+
+    public void setAutoRefresh(final boolean enabled)
+    {
+        if ( enabled )
+        {
+            adView.startAutoRefresh();
+        }
+        else
+        {
+            adView.stopAutoRefresh();
+        }
+    }
+
+    public void setExtraParameters(@Nullable final Map<String, Object> extraParameters)
+    {
+        if ( extraParameters == null ) return;
+
+        for ( Map.Entry<String, Object> entry : extraParameters.entrySet() )
+        {
+            adView.setExtraParameter( entry.getKey(), (String) entry.getValue() );
+        }
+    }
+
+    public void setLocalExtraParameters(@Nullable final Map<String, Object> localExtraParameters)
+    {
+        if ( localExtraParameters == null ) return;
+
+        for ( Map.Entry<String, Object> entry : localExtraParameters.entrySet() )
+        {
+            adView.setLocalExtraParameter( entry.getKey(), entry.getValue() );
+        }
+    }
+
+    public boolean hasContainerView()
+    {
+        return containerView != null;
+    }
+
+    public void attachAdView(AppLovinMAXAdView view)
+    {
+        containerView = view;
+    }
+
+    public void detachAdView()
+    {
+        containerView = null;
+    }
+
+    public void loadAd()
+    {
+        adView.loadAd();
+    }
+
+    public void destroy()
+    {
+        detachAdView();
+
+        adView.setListener( null );
+        adView.setRevenueListener( null );
+        adView.destroy();
+    }
+
+    @Override
+    public void onAdLoaded(@NonNull final MaxAd ad)
+    {
+        Map<String, Object> params = AppLovinMAX.getInstance().getAdInfo( ad );
+
+        if ( isPreloadPlatformWidget )
+        {
+            AppLovinMAX.getInstance().fireCallback( "OnPlatformWidgetAdViewAdLoadedEvent", Map.copyOf( params ) );
```

Comment:
I think there was a reason why we had to copy the params, maybe worth adding a comment for it in case we forget years from now?